### ISSUE TITLE
[Feature Request] Avoid pickling object attributes matching user-defined predicates

### ISSUE THREAD

**Issue by leycec** [uqfoundation/dill#225]
Greetings from the still-frigid Great White North (i.e., _Canada_). Upfront apologies for polluting your issue tracker with yet another frivolous feature request: in this case, **global attribute filtering.**

## What U Want?

Our [downstream application](https://gitlab.com/betse/betse) would *love* to avoid pickling any object attributes whose attribute names match a user-defined predicate – regardless of the types of those objects. Sadly, neither the underlying `pickle` nor higher-level `dill` APIs appear to be extensible in this way. Both permit callers to define custom reduction functions uniquely applicable to *specific* types but appear to have no means of defining custom reduction functions globally applicable to *all* types. 

## Why U Want This?

A bit of context might be useful. Like most ~~obese~~ large-scale Python projects, ours internally defines a [`@property_cached` decorator](https://gitlab.com/betse/betse/blob/master/betse/util/type/call/memoizers.py) for efficiently caching property return values into object attributes. The names of these attributes are *all* prefixed by the magic substring `__property_cached_`.

Efficiency is fun, so this works. **Except under pickling.**

These attributes tend to be quite large in worst-case scenarios – like several gigabytes too large. Since these attributes are all safely recomputable at runtime, we'd rather not pickle them if we don't have to. Since any arbitrary object of any arbitrary type may contain methods decorated by `@property_cached`, attempting to exclude these attributes from pickling by defining custom reduction functions specific to particular types fails to scale.

A general-purpose strategy is needed.

## Why U No Do This the Easy Way?

"It'll be easy!", I initially said. It wasn't easy. After several hours of futile grepping through the `pickle` and `dill` codebases, I'm no closer to a sane solution. At this point, I'll happily accept an insane solution.

The only one on hand appears to be:

* Define a new application specific `SmartPickleableMixin` class implementing the `__getstate__()` special method to omit all object attributes prefixed by the magic substring `__property_cached_`.
* Refactor *all* existing application classes to additionally subclass the `SmartPickleableMixin` class.

This is pretty insane for any number of obvious reasons, the least of which is that implementing this change across a codebase exceeding 100,000 lines of raw, brutal, no-holds-barred scientific Python will probably require two human lifetimes. Moreover, it's manual and hence fragile – really, *really* fragile.

Humans are imperfect meat bags. Any class that I or other project contributors accidentally neglect to subclass from `SmartPickleableMixin` will continue to silently pickle cached properties – which is bad.

The younger me might try monkey-patching that `__getstate__()`  implementation into the root `object` type, transparently enabling this functionality across the entire codebase. The older me merely squints suspiciously and refuses to do anything.

Hence, this modest request for a global attribute filtering solution applicable to all types. Maybe this already exists? My grep-fu isn't what it was. <sup>_Damn these rheumy eyes!_</sup>

## This Issue Too Long

Right. Sorry about that.

**In summary:** `dill` is simply amazing. It's single-handedly made our scientific use case feasible. Whether this feature ever comes to fruition or not, thanks for this laborious work of gritty, low-level volunteerism.

Springtime pineapples for all! :pineapple: :pineapple: :pineapple:

**Comment by matsjoyce** [comment]
+1, I like funny issues! Try this:

```py3
# Modify pickler

import dill

class CustomPickler(dill.Pickler):
    def save(self, obj, *args, **kwargs):
        # Filter out cached attrs
        tmp_rm = {}
        if hasattr(obj, "__dict__"):
            for key in list(obj.__dict__):
                if type(key) is str and key.startswith("__property_cached_"):
                    tmp_rm[key] = obj.__dict__.pop(key)
        # Pickle obj
        super().save(obj, *args, **kwargs)
        # Restore cached attrs
        for key, value in tmp_rm.items():
            obj.__dict__[key] = value

dill.dill.Pickler = CustomPickler

# Testing

from functools import wraps

def property_cached(property_method):
    func_body = '''
@wraps(__property_method)
def property_method_cached(self, __property_method=__property_method):
    try:
        return self.{property_name}
    except AttributeError:
        self.{property_name} = __property_method(self)
        return self.{property_name}
'''.format(property_name='__property_cached_' + property_method.__name__)
    local_attrs = {'__property_method': property_method}
    exec(func_body, globals(), local_attrs)
    return property(local_attrs['property_method_cached'])


class A:
    def __init__(self, x, y, z):
        self.x, self.y, self.z = x, y, z

    @property_cached
    def magic(self):
        return "a" * 10000

a=A(1, None, set())
assert a.magic == "a" * 10000
assert a.magic == "a" * 10000

raw = dill.dumps(a)
assert len(raw) < 10000

assert a.__property_cached_magic == "a" * 10000

clone = dill.loads(raw)
assert clone.x == 1
assert clone.y is None
assert clone.z == set()
assert clone.magic == "a" * 10000

```
It should filter out all attributes beginning with `__property_cached_`.

**Comment by leycec** [comment]
**Oh, man.** Like a half-starved triathlete subsisting on vegan nut butter and dreams, you went the distance. You delivered a full-bodied solution integrating our half-crazed application-specific decorator – complete with formalized unit tests that actually all pass.

It pays to be punny.

## Feature Request Part II: The Legend of Brutal Pain Continues

You're probably going to hate me for this, but my only question now is: *"Now how do I do that recursively over the entire object graph of objects reachable from the root object to be pickled?"* 

Actually, even that was a lie. I have a second question, which is even worse: *"Now how do I make that space- and time-performant?"*

While recursively traversing the object graph certainly seems feasible, preserving and restoring filtered attributes for all objects in that graph seems... *costly.* A two-dimensional `dict` of `dict`s mapping from object IDs to `dict`s mapping from removed attribute names to values would probably do it, but... **urgh.**

Actually, scratch that. In our case, cached properties are all safely deletable; on the next access, they'll simply be transparently recreated. So, the recursive traversal can be simplified by directly deleting all object attributes whose names are prefixed by `__property_cached_` without attempting to preserve and restore those attributes. *Phew!*

That's nice. But I'm still scared. This has become the Mount Doom of `dill` feature requests.

Please don't feel obligated to cobble together another awesome solution. Time is precious. I've consumed enough of yours, already.

## Feature Request Part III: A Whole New World

The fact that I'm now attempting to perform a recursive breadth- or depth-first visitation of every node of an object graph merely to avoid pickling cached attributes tells me I'm squarely in FUBAR Land™. I'm getting that itching, scabrous feeling on my neck again. Something has gone horribly wrong.

There *must* be a better way. My neck deserves better.

Fortunately, `pickle` and `dill` already provide an out-of-the-box solution for filtering attributes: the type-specific `__getstate__()` protocol.  The dwindling voice of reason inside me suggests that I should probably just work within the confines of that elegant API rather than attempting to brute-force recursive traversal into a `Pickler` replacement.<sup>_...as much fun as that would surely be. Don't get me wrong here._</sup>

In short, I'm considering refactoring our `@property_cached` decorator. Currently, this decorator decorates the passed function with *another* function caching properties onto the parent object directly. That doesn't help us. The "...caching properties onto the parent object directly" part is the problem. We can't reasonably implement a universal `__getstate__()` method filtering out cached properties from all possible objects.

At this point, your head is probably planted firmly onto the keyboard. Drool and spittle are beginning to collect in the adjustable ergonomic keyboard tray. But what if it didn't have to be this way?

What if this decorator instead decorated the passed function with a callable class implementing the `__call__()`, `__get__()`, and `__getstate__()` protocols to cache properties onto *itself* rather than the parent object? Specifically:

* `__call__()` will cache the value returned by the decorated function onto the bound method returned by `__get__()`.
* `__get__()` will create and return a bound method, satisfying the descriptor protocol also implemented by the `property` descriptor. Hence, we won't need to decorate this callable class with `property`. <sup>_That's nice, dear._</sup>
*  `__getstate__()` will exclude the value cached onto that bound method. *Hmmm.* **Wait.** That means that `__getstate__()` needs to be defined on the bound method instead, doesn't it? Jesus-on-chocolate-stick, preserve us. I suppose I could define a new `MethodType` subclass implementing `__getstate__()` to exclude the value cached onto itself, but... *urgh.*

I'm perfectly confident that could maybe work.

**Comment by matsjoyce** [comment]
The `save` method is already recursively called (this is how pickle works):

```py3
a=A(1, None, A(1, None, A(1, None, set())))
assert a.magic == "a" * 10000
assert a.z.magic == "a" * 10000

raw = dill.dumps(a)
assert len(raw) < 10000

assert a.__property_cached_magic == "a" * 10000

clone = dill.loads(raw)
assert clone.x == 1
assert clone.y is None
assert type(clone.z) is A
assert clone.z.x == 1
assert clone.z.y is None
assert type(clone.z.z) is A
assert clone.z.z.x == 1
assert clone.z.z.y is None
assert clone.z.z.z == set()
assert clone.magic == "a" * 10000
```
Me no see no problem?

**Comment by leycec** [comment]
**Jeebuz.** Well, *that's* a clever bit of API design. The Visitor pattern strikes back. I missed that watershed realization in my speed-read over the `dill` and `pickle` documentation.

I feel doubly bad. Again, thanks for supplying a brilliant self-documenting solution. I'll patch this into our [biophysics simulator](https://gitlab.com/betse/betse) shortly – with full credits to you. Thus concludes the sordid tale of this feature request.

Also, *Battle for Wesnoth* rocketh. Your contributions there are appreciated. That is all. :shield:

**Comment by leycec** [comment]
[**It is done,**](https://gitlab.com/betse/betse/blob/master/betse/lib/pickle/pickles.py) credits and all. Thanks yet again. You saved all our disk space.
